sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -508,7 +508,17 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       sparkSession.sessionState.conf.sessionLocalTimeZone)
     val filteredLines: Dataset[String] =
       CSVUtils.filterCommentAndEmpty(csvDataset, parsedOptions)
-    val maybeFirstLine: Option[String] = filteredLines.take(1).headOption
+
+    // For performance, short-circuit the collection of the first line when it won't be used:
+    //   - TextInputCSVDataSource - Only uses firstLine to infer an unspecified schema
+    //   - CSVHeaderChecker       - Only uses firstLine to check header, when headerFlag is true
+    //   - CSVUtils               - Only uses firstLine to filter headers, when headerFlag is true
```

Comment:
Would you mind adding a note that says if this logic is getting complicated, it might have to better refactor?

Reply:
Yes, good idea. Just expanded the inline comment to cover this. Hope it adequately conveys the point. Let me know if you have any suggestions or prefer different phrasing.